### PR TITLE
feat (worker): worker download

### PR DESCRIPTION
**Pull request by yesnault**
Signed-off-by: Yvonnick Esnault <yvonnick.esnault@corp.ovh.com>

### CODE REVIEW

**Comment by yesnault** [comment]
cc @blaryjp

File: engine/worker/cmd_download.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+		req, errRequest := http.NewRequest("POST", fmt.Sprintf("http://127.0.0.1:%d/download", port), bytes.NewReader(data))
+		if errRequest != nil {
+			sdk.Exit("cannot post worker download (Request): %s\n", errRequest)
+		}
+
+		client := http.DefaultClient
+		client.Timeout = 5 * time.Minute
+
+		resp, errDo := client.Do(req)
+		if errDo != nil {
```
Comment (fsamin) [inline engine/worker/cmd_download.go]: what about try to decode the business error ? 
example in cmd_tmpl.go

File: engine/worker/cmd_download.go [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+			sdk.Exit("cannot artefact download HTTP %d\n", resp.StatusCode)
+		}
+	}
+}
+
+func (wk *currentWorker) downloadHandler(w http.ResponseWriter, r *http.Request) {
+	// Get body
+	data, errRead := ioutil.ReadAll(r.Body)
+	if errRead != nil {
+		w.WriteHeader(http.StatusBadRequest)
```
Comment (fsamin) [inline engine/worker/cmd_download.go]:
```golang
newError := sdk.NewError(sdk.ErrWrongRequest, errRead)
writeError(w, r, newError)
return
```

File: engine/worker/cmd_download.go [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+		return
+	}
+
+	var reqArgs workerDownloadArtifact
+	if err := json.Unmarshal(data, &reqArgs); err != nil {
+		w.WriteHeader(http.StatusBadRequest)
```
Comment (fsamin) [inline engine/worker/cmd_download.go]: writeError

File: engine/worker/cmd_download.go [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+		return
+	}
+
+	sendLog := getLogger(wk, wk.currentJob.pbJob.ID, wk.currentJob.currentStep)
+
+	if wk.currentJob.wJob == nil {
+		sendLog("command 'worker download' is only available on CDS Workflows")
+		w.WriteHeader(http.StatusBadRequest)
```
Comment (fsamin) [inline engine/worker/cmd_download.go]: writeError

File: engine/worker/cmd_download.go [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+			sdk.Exit("internal error (%s)\n", errMarshal)
+		}
+
+		req, errRequest := http.NewRequest("POST", fmt.Sprintf("http://127.0.0.1:%d/download", port), bytes.NewReader(data))
+		if errRequest != nil {
+			sdk.Exit("cannot post worker download (Request): %s\n", errRequest)
+		}
+
+		client := http.DefaultClient
+		client.Timeout = 5 * time.Minute
```
Comment (fsamin) [inline engine/worker/cmd_download.go]: I'm afraid 5 minutes could be too short if there is bunch of big artifacts

Comment (bnjjj) [inline engine/worker/cmd_download.go]: So if the download is longer than 5 minutes it will be stopped ?

File: engine/worker/cmd_download.go [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+		return
+	}
+
+	var reqArgs workerDownloadArtifact
+	if err := json.Unmarshal(data, &reqArgs); err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	sendLog := getLogger(wk, wk.currentJob.pbJob.ID, wk.currentJob.currentStep)
```
Comment (fsamin) [inline engine/worker/cmd_download.go]: usually we don't send logs, we return `sdk.Error` and the worker send the logs

File: engine/worker/cmd_download.go [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+		reqArgs.Workflow = sdk.ParameterValue(wk.currentJob.params, "cds.workflow")
+	}
+
+	if reqArgs.Number == 0 {
+		var errN error
+		buildNumberString := sdk.ParameterValue(wk.currentJob.params, "cds.run.number")
+		reqArgs.Number, errN = strconv.ParseInt(buildNumberString, 10, 64)
+		if errN != nil {
+			sendLog(fmt.Sprintf("Cannot parse '%s' as run number: %s", buildNumberString, errN))
+			w.WriteHeader(http.StatusBadRequest)
```
Comment (fsamin) [inline engine/worker/cmd_download.go]: writeError

File: engine/worker/cmd_download.go [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+			return
+		}
+	}
+
+	projectKey := sdk.ParameterValue(wk.currentJob.params, "cds.project")
+	artifacts, err := wk.client.WorkflowRunArtifacts(projectKey, reqArgs.Workflow, reqArgs.Number)
+	if err != nil {
+		sendLog(fmt.Sprintf("Cannot download artifacts with worker download: %s", err))
+		w.WriteHeader(http.StatusBadRequest)
```
Comment (fsamin) [inline engine/worker/cmd_download.go]: writeError

File: engine/worker/cmd_download.go [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+		return
+	}
+
+	regexp, errp := regexp.Compile(reqArgs.Pattern)
+	if errp != nil {
+		sendLog(fmt.Sprintf("Invalid pattern %s : %s", reqArgs.Pattern, errp))
+		w.WriteHeader(http.StatusBadRequest)
```
Comment (fsamin) [inline engine/worker/cmd_download.go]: writeError

File: engine/worker/cmd_download.go [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+			wg.Done()
+			continue
+		}
+
+		go func(a *sdk.WorkflowNodeRunArtifact) {
+			defer wg.Done()
+
+			f, err := os.OpenFile(a.Name, os.O_RDWR|os.O_CREATE, os.FileMode(a.Perm))
+			if err != nil {
+				sendLog(fmt.Sprintf("Cannot download artifact (OpenFile) %s: %s", a.Name, err))
```
Comment (fsamin) [inline engine/worker/cmd_download.go]: return the error

File: engine/worker/cmd_download.go [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+				return
+			}
+			sendLog(fmt.Sprintf("downloading artifact %s from workflow %s/%s on run %d...", a.Name, projectKey, reqArgs.Workflow, reqArgs.Number))
+			if err := wk.client.WorkflowNodeRunArtifactDownload(projectKey, reqArgs.Workflow, *a, f); err != nil {
+				sendLog(fmt.Sprintf("Cannot download artifact %s: %s", a.Name, err))
```
Comment (fsamin) [inline engine/worker/cmd_download.go]: return the error

File: engine/worker/cmd_download.go [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+	regexp, errp := regexp.Compile(reqArgs.Pattern)
+	if errp != nil {
+		sendLog(fmt.Sprintf("Invalid pattern %s : %s", reqArgs.Pattern, errp))
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	wg := new(sync.WaitGroup)
+	wg.Add(len(artifacts))
+
+	for i := range artifacts {
```
Comment (fsamin) [inline engine/worker/cmd_download.go]: this is kind of duplicate code

File: engine/worker/cmd_download.go (hunk 10 above)
Comment (yesnault) [reply engine/worker/cmd_download.go]: done

File: engine/worker/cmd_download.go (hunk 11 above)
Comment (yesnault) [reply engine/worker/cmd_download.go]: done

File: engine/worker/cmd_download.go [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+
+		client := http.DefaultClient
+
+		resp, errDo := client.Do(req)
+		if errDo != nil {
+			sdk.Exit("cannot post worker download (Do): %s\n", errDo)
+		}
+
+		if resp.StatusCode >= 300 {
+			body, err := ioutil.ReadAll(resp.Body)
```
Comment (fsamin) [inline engine/worker/cmd_download.go]: close the body

File: engine/worker/cmd_download.go [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+			}
+			cdsError := sdk.DecodeError(body)
+			sdk.Exit("download failed: %v\n", cdsError)
+		}
+	}
+}
+
+func (wk *currentWorker) downloadHandler(w http.ResponseWriter, r *http.Request) {
+	// Get body
+	data, errRead := ioutil.ReadAll(r.Body)
```
Comment (fsamin) [inline engine/worker/cmd_download.go]: close the body

File: engine/worker/cmd_download.go [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+	regexp, errp := regexp.Compile(reqArgs.Pattern)
+	if errp != nil {
+		newError := sdk.NewError(sdk.ErrWrongRequest, fmt.Errorf("Invalid pattern %s : %s", reqArgs.Pattern, errp))
+		writeError(w, r, newError)
+		return
+	}
+	wg := new(sync.WaitGroup)
+	wg.Add(len(artifacts))
+
+	var isInError bool
```
Comment (fsamin) [inline engine/worker/cmd_download.go]: should use a multiError to concat all the errors

File: engine/worker/cmd_download.go (hunk 11 above)
Comment (fsamin) [reply engine/worker/cmd_download.go]: should not log the error

File: engine/worker/cmd_download.go [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+				return
+			}
+			sendLog(fmt.Sprintf("downloading artifact %s from workflow %s/%s on run %d...", a.Name, projectKey, reqArgs.Workflow, reqArgs.Number))
+			if err := wk.client.WorkflowNodeRunArtifactDownload(projectKey, reqArgs.Workflow, *a, f); err != nil {
+				sendLog(fmt.Sprintf("Cannot download artifact %s: %s", a.Name, err))
+				isInError = true
+				return
+			}
+			if err := f.Close(); err != nil {
+				sendLog(fmt.Sprintf("Cannot download artifact %s: %s", a.Name, err))
```
Comment (fsamin) [inline engine/worker/cmd_download.go]: should not log the error

**Pull request merged**